doc/proposals/images-subspec.md

```diff
@@ -0,0 +1,190 @@
+form of pods, deployments or other kubernetes resources is not always clearly
+defined. This proposal would require that images would be exposed and
+overridable on the operator custom resource definitions (CRDs). This allows a user of a deployed
+operator to override the images needed at deployment of the custom resource
+(CR).
+
+## Motivation
+
+This proposal is driven by a few motivations:
+
+- Makes operators more transparent with what is being installed on the cluster.
+- Formalizes a best practice of operator development.
+- The depencency between operator version and operator resource version is broken.
+- Allows for out of band updates of the underlying images.
+- Registry information for the images can easily be overridden to point at
+  private registries.
+- Provides some standardization to the operator CRD pattern.
+
+## Goals
+
+- Provide an easy mechanism of defining and overriding images and related image
+  information (like pull secrets).
+- Ensure the operators respect overridden images.
+- Increase transparency with what operators are installing.
+
+### Non-Goals
+
+- OLM management of the images spec is not in this proposal.
+- Mapping/re-mapping other container values (ports, etc.).
+
+## Proposal
+
+Implementing the proposal would include:
+
+- Add an optional section to the CRD spec named images that contains definitions of
+  all images used by the operator. An example:
+
+  ```yaml
+  apiVersion: app.example.com/v1alpha1
+  kind: AppService
+  metadata:
+    name: example-appservice
+  spec:
+    size: 3
+    images:
+      - name: nginx
+        image: nginx:1.7.9
+      - name: example
+        iamge: my-registry:5000/example:1.0.0
+  ```
+
+- The images spec property is required to be read by the operator to deploy resources
+  it creates (pods, deployments, statefulsets) where additional images are used.
+- Add scorecard tests to verify the images spec is being used.
+- Modify any operator-sdk clients to generate the images spec.
+
+### User Stories
+
+#### Add a spec section named images under CRD properties.
+
+Create a custom spec interface that the generated spec objects will extend.
+The spec interface includes a new list of objects called `images`. These images follows the
+OpenAPIV3 spec listed below. Additionally add APIs that transform images rows to
+Kubernetes API resources to help developers onboard.
+
+```yaml
+images:
+  type: array
+  items:
+    type: object
+    properties:
+      name:
+        type: string
+        description: Name of image used for lookup.
+      image:
+        type: string
+        description: Location of the image.
+      pullPolicy:
+        type: string
+        description: Pull Policy for the image.
+      pullSecrets:
+        type: array
+        description: Array of pull secret names to use.
+        items:
+          types: string
+    required: ['name', 'image']
+```
+
+#### Add a scorecard test to verify changes to image locations.
+
```

Comment:
I'd keep just the impl to gen the basic code and keep a smaller scope. If it is accepted and implemented, then we can have a better idea over how should be the next steps or additional needs.

Reply:
Sure.